src/FNA3D_Driver_Vulkan.c

```diff
@@ -2892,65 +3108,284 @@ static uint8_t VULKAN_INTERNAL_AllocateMemory(
 
-	/* find the largest free region and use it */
-	if (allocator->sortedFreeRegionCount > 0)
+	if (allocationIndex != allocator->allocationCount - 1)
 	{
-		region = allocator->sortedFreeRegions[0];
-		allocation = region->allocation;
+		allocator->allocations[allocationIndex] = allocator->allocations[allocator->allocationCount - 1];
+	}
 
-		alignedOffset = VULKAN_INTERNAL_NextHighestAlignment(
-			region->offset,
-			memoryRequirements->memoryRequirements.alignment
-		);
+	allocator->allocationCount -= 1;
+}
 
-		if (alignedOffset + requiredSize <= region->offset + region->size)
-		{
-			*pMemoryAllocation = allocation;
+static uint8_t VULKAN_INTERNAL_AllocateMemory(
+	VulkanRenderer *renderer,
+	VkBuffer buffer,
+	VkImage image,
+	uint32_t memoryTypeIndex,
+	VkDeviceSize allocationSize,
+	uint8_t dedicated,
+	uint8_t isHostVisible,
+	VulkanMemoryAllocation **pMemoryAllocation
+) {
+	VulkanMemoryAllocation *allocation;
+	VulkanMemorySubAllocator *allocator = &renderer->memoryAllocator->subAllocators[memoryTypeIndex];
+	VkMemoryAllocateInfo allocInfo;
+	VkMemoryDedicatedAllocateInfoKHR dedicatedInfo;
+	VkResult result;
 
-			/* not aligned - create a new free region */
-			if (region->offset != alignedOffset)
-			{
-				VULKAN_INTERNAL_NewMemoryFreeRegion(
-					allocation,
-					region->offset,
-					alignedOffset - region->offset
-				);
-			}
+	allocInfo.sType = VK_STRUCTURE_TYPE_MEMORY_ALLOCATE_INFO;
+	allocInfo.memoryTypeIndex = memoryTypeIndex;
+	allocInfo.allocationSize = allocationSize;
 
-			*pOffset = alignedOffset;
-			*pSize = requiredSize;
+	allocation = SDL_malloc(sizeof(VulkanMemoryAllocation));
+	allocation->size = allocationSize;
+	allocation->mapLock = SDL_CreateMutex();
+
+	if (dedicated)
+	{
+		dedicatedInfo.sType =
+			VK_STRUCTURE_TYPE_MEMORY_DEDICATED_ALLOCATE_INFO_KHR;
+		dedicatedInfo.pNext = NULL;
+		dedicatedInfo.buffer = buffer;
+		dedicatedInfo.image = image;
+
+		allocInfo.pNext = &dedicatedInfo;
+
+		allocation->dedicated = 1;
+	}
+	else
+	{
+		allocInfo.pNext = NULL;
+
+		allocator->allocationCount += 1;
+		allocator->allocations = SDL_realloc(
+			allocator->allocations,
+			sizeof(VulkanMemoryAllocation*) * allocator->allocationCount
+		);
+
+		allocator->allocations[
+			allocator->allocationCount - 1
+		] = allocation;
+
+		allocation->dedicated = 0;
+	}
+
+	allocation->usedRegions = SDL_malloc(sizeof(VulkanMemoryUsedRegion*));
+	allocation->usedRegionCount = 0;
+	allocation->usedRegionCapacity = 1;
+
+	allocation->freeRegions = SDL_malloc(sizeof(VulkanMemoryFreeRegion*));
+	allocation->freeRegionCount = 0;
+	allocation->freeRegionCapacity = 1;
+
+	allocation->allocator = allocator;
+
+	result = renderer->vkAllocateMemory(
+		renderer->logicalDevice,
+		&allocInfo,
+		NULL,
+		&allocation->memory
+	);
+
+	if (result != VK_SUCCESS)
+	{
+		/* Uh oh, we couldn't allocate, time to clean up */
+		SDL_free(allocation->freeRegions);
+
+		allocator->allocationCount -= 1;
+		allocator->allocations = SDL_realloc(
```

Comment:
Is this really wise? It could force a new copy of the allocations buffer to be allocated (1 element smaller) and then copied into, which is questionable if we're already low on memory

Reply:
IMO if the application is so memory-starved that a realloc fails we are already extremely hosed